pymc3/distributions/simulator.py

```diff
@@ -14,127 +14,314 @@
 
-__all__ = ["Simulator"]
+__all__ = ["Simulator", "SimulatorRV"]
 
 _log = logging.getLogger("pymc3")
 
 
+def identity(x):
+    """Identity function, used as a summary statistics."""
+    return x
+
+
+def gaussian(epsilon, obs_data, sim_data):
+    """Gaussian kernel."""
+    return -0.5 * ((obs_data - sim_data) / epsilon) ** 2
+
+
+def laplace(epsilon, obs_data, sim_data):
+    """Laplace kernel."""
+    return -at.abs_((obs_data - sim_data) / epsilon)
+
+
+class SimulatorRV(RandomVariable):
+    """
+    Base class for SimulatorRVs
+
+    This should be subclassed when defining custom Simulator objects.
+
+    The following class attributes must be defined:
+
+    ndim_supp : int
+        Number of dimensions of the SimulatorRV (0 for scalar, 1 for vector, etc.)
+    ndims_params: list[int]
+        Number of dimensions of each parameter of the RV, plus a 0 for the
+        epsilon parameter. For example, if the Simulator accepts two scalar
+        inputs, it should be `(0, 0, 0)`
+    fn: callable
+        Python random simulator function. Should expect the following signature
+        (rng, arg1, arg2, ... argn, size), where rng is a numpy.random.RandomStream()
+        and size defines the size of the desired sample. The epsilon parameter
+        is not passed to the callable.
+
+    The following class attributes can be optionally defined:
+
+    distance : Aesara Op, callable or str
+        Distance function. Available options are `"gaussian"` (default), `"laplacian"`,
+        `"kullback_leibler"` or a user defined function (or Aesara Op) that takes
+        epsilon, the summary statistics of observed_data and the summary statistics
+        of simulated_data as input.
+        ``gaussian`` :math: `-0.5 \\left(\\left(\frac{xo - xs}{\\epsilon}\right)^2\right)`
+        ``laplace`` :math: `{\\left(\frac{|xo - xs|}{\\epsilon}\right)}`
+        ``kullback_leibler`` `:math: d \\sum(-\\log(\frac{\nu_d} {\rho_d}) / \\epsilon) + log_r`
+        gaussian + ``sum_stat="sort"`` is equivalent to the 1D 2-wasserstein distance
+        laplace + ``sum_stat="sort"`` is equivalent to the the 1D 1-wasserstein distance
+
+    sum_stat: Aesara Op, callable or str
+        Summary statistic function. Available options are `"indentity"` (default),
+        `"sort"`, `"mean"`, `"median"`. If a callable (or Aesara Op) is defined,
+        it should return a 1d numpy array (or Aesara vector).
+
+    Other class attributes used by `RandomVariables` can also be defined. See the
+    documentation for aesara.tensor.random.op.RandomVariable for more information.
+
+    Examples
+    --------
+    .. code-block:: python
+
+        def my_simulator_fn(rng, loc, scale, size):
+            return rng.normal(loc, scale, size=size)
+
+        class MySimulatorRV(pm.SimulatorRV):
+            ndim_supp = 0
+            ndims_params = [0, 0, 0]
+            fn = my_simulator_fn
+            distance = "gaussian"
+            sum_stat = "sort"
```

Comment:
Shouldn't the name be implemented by a subclass too?

Reply:
It can be implemented optionally, but I did not want to make it a requirement if it's not needed